### PR TITLE
Fix type issues

### PR DESCRIPTION
**Pull request by muglug**
After seeing the work done in https://github.com/composer/composer/pull/7883, I ran Psalm on the repo.

The only obvious bug it found was mentioned in https://github.com/composer/composer/commit/902cb290e792fdc3a4ff6a7af23ce7fea54b7acc#r32358863, but running it highlighted some redundant code and invalid docblocks.

I can create a separate PR to add Psalm to the build, if you'd like - it would look like this: https://github.com/muglug/composer/pull/2/commits/d14d23949152ea940457e48b7bb17b59a5c9047b.

### CODE REVIEW

**Comment by Seldaek** [comment]
I am not sure what to do here.. Skimming through changes I spotted a few which are incorrect, which makes me think I'd have to go through all of them in depth before merging. There are definitely some correct ones too though so closing outright isn't good either.

Why are some nullables switched to `?foo` while others are `foo|null`? I believe the latter is preferred as some tooling doesn't support `?foo` in phpdoc.

Adding Psalm to the build I'd rather not as it'll probably be quite redundant with PHPStan if we increase the check level there. It's also one more error-exclusion list to maintain.

File: src/Composer/Command/DiagnoseCommand.php [hunk 1]
```diff
@@ -602,20 +602,6 @@ private function checkPlatform()
                         $text .= "Install either of them or recompile php without --disable-iconv";
                         break;
 
-                    case 'unicode':
```
Comment (staabm) [inline src/Composer/Command/DiagnoseCommand.php]: this changes seem unrelated?

Comment (muglug) [reply src/Composer/Command/DiagnoseCommand.php]: Oh it’s not - if you look at the code above this condition can never be met.

File: src/Composer/Util/RemoteFilesystem.php [hunk 2]
```diff
@@ -82,7 +82,7 @@ public function __construct(IOInterface $io, Config $config, array $options = ar
      * @param bool   $progress  Display the progression
      * @param array  $options   Additional context options
      *
-     * @return bool true
+     * @return bool|string true
```
Comment (Seldaek) [inline src/Composer/Util/RemoteFilesystem.php]: This change is invalid. get() will return bool when $fileName is passed in.

File: src/Composer/Util/Loop.php [hunk 3]
```diff
@@ -29,6 +29,7 @@ public function __construct(HttpDownloader $httpDownloader)
 
     public function wait(array $promises)
     {
+        /** @var ?\Exception */
```
Comment (Seldaek) [inline src/Composer/Util/Loop.php]: Still some `?` nullable here :)

File: src/Composer/Util/Bitbucket.php [hunk 4]
```diff
@@ -202,6 +202,7 @@ public function requestToken($originUrl, $consumerKey, $consumerSecret)
 
         $this->storeInAuthConfig($originUrl, $consumerKey, $consumerSecret);
 
+        /** @var array $this->token */
```
Comment (Seldaek) [inline src/Composer/Util/Bitbucket.php]: This seems kinda useless to me. The tool might need it, but that seems like a bug in the tool then. The property has a type declaration already. Please remove.

File: src/Composer/IO/IOInterface.php [hunk 5]
```diff
@@ -160,7 +160,7 @@ public function askAndHideAnswer($question);
      * @param bool        $multiselect  Select more than one value separated by comma
      *
      * @throws \InvalidArgumentException
-     * @return int|string|array          The selected value or values (the key of the choices array)
+     * @return int|string|array|bool    The selected value or values (the key of the choices array)
```
Comment (Seldaek) [inline src/Composer/IO/IOInterface.php]: This probably sohuld be `mixed` really at this point ;)

File: src/Composer/Util/Bitbucket.php (hunk 4 above)
Comment (muglug) [reply src/Composer/Util/Bitbucket.php]: yeah, I agree. It was added to reset the tool's knowledge of `$this->token`, which already had a !empty check above. I'll think about how this could be dealt with better.

File: src/Composer/Downloader/DownloadManager.php [hunk 6]
```diff
@@ -294,9 +294,7 @@ public function update(PackageInterface $initial, PackageInterface $target, $tar
 
         // if downloader type changed, or update failed and user asks for reinstall,
         // we wipe the dir and do a new install instead of updating it
-        if ($initialDownloader) {
-            $initialDownloader->remove($initial, $targetDir);
-        }
+        $initialDownloader->remove($initial, $targetDir);
```
Comment (Seldaek) [inline src/Composer/Downloader/DownloadManager.php]: This introduces a bug. $initialDownloader might be null as getDownloaderForPackage can return null. I am not sure why this change was done.

Comment (muglug) [reply src/Composer/Downloader/DownloadManager.php]: It does not, as far as I can see.

Have a look at the code above:

```
        if (!$initialDownloader && !$downloader) {
            return;
        }
        // if we have a downloader present before, but not after, the package became a metapackage and its files should be removed
        if (!$downloader) {
            $initialDownloader->remove($initial, $targetDir);
            return;
        }
```

After that, it's impossible for `$initialDownloader` to be null. In fact it's assumed to be non-null when used in `$this->getDownloaderType($initialDownloader)`.

This is one of the areas where Psalm is better than PHPStan - it includes some (very basic) SAT solving to help it understand the above logic.

File: src/Composer/DependencyResolver/GenericRule.php [hunk 7]
```diff
@@ -23,10 +23,10 @@ class GenericRule extends Rule
     protected $literals;
 
     /**
-     * @param array                 $literals
-     * @param int                   $reason     A RULE_* constant describing the reason for generating this rule
-     * @param Link|PackageInterface $reasonData
-     * @param array                 $job        The job this rule was created from
+     * @param array                      $literals
+     * @param int|null                   $reason     A RULE_* constant describing the reason for generating this rule
+     * @param Link|PackageInterface|null $reasonData
```
Comment (Seldaek) [inline src/Composer/DependencyResolver/GenericRule.php]: Also not sure why those two were marked nullable. Rather undo unless there's a good explanation.

File: src/Composer/DependencyResolver/Solver.php [hunk 8]
```diff
@@ -513,7 +514,7 @@ protected function analyze($level, Rule $rule)
             );
         }
 
-        $newRule = new GenericRule($learnedLiterals, Rule::RULE_LEARNED, $why);
+        $newRule = new GenericRule($learnedLiterals, Rule::RULE_LEARNED, null);
```
[the remark this answers is not in the document]
Comment (muglug) [reply src/Composer/DependencyResolver/Solver.php]: The third arg of `GenericRule::__construct` expects `Composer\Package\PackageInterface|Composer\Package\Link`, but `$why` is an `int`.

File: src/Composer/DependencyResolver/GenericRule.php (hunk 7 above)
Comment (muglug) [reply src/Composer/DependencyResolver/GenericRule.php]: You pass `null` in as the third argument in a number of places (happy to provide examples) so I thought it should also be in the signature.

File: src/Composer/Command/ConfigCommand.php [hunk 9]
```diff
@@ -226,7 +226,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
         }
 
         $settingKey = $input->getArgument('setting-key');
-        if (!$settingKey) {
+        if (!$settingKey || !is_string($settingKey)) {
```
Comment (Seldaek) [inline src/Composer/Command/ConfigCommand.php]: This adds no value AFAICT.

File: src/Composer/Command/BaseCommand.php [hunk 10]
```diff
@@ -27,6 +27,8 @@
 /**
  * Base class for Composer commands
  *
+ * @method Application getApplication()
```
Comment (Seldaek) [inline src/Composer/Command/BaseCommand.php]: This shouldn't be needed, it's defined on the parent class.

File: src/Composer/DependencyResolver/GenericRule.php (hunk 7 above)
Comment (Seldaek) [reply src/Composer/DependencyResolver/GenericRule.php]: Ok in that case fair enough leave it.

File: src/Composer/DependencyResolver/Solver.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (Seldaek) [reply src/Composer/DependencyResolver/Solver.php]: Right, but in Rule.php:240 it appears to use it as an int, so IMO the doctype should be expanded rather than removing data here.

Generally speaking when following static analysis output I'd tend to suspect the docs of being out of date rather than the code. Fixing the code can lead to bugs, fixing docs can only lead in confusion at worst.

File: src/Composer/Command/ConfigCommand.php (hunk 9 above)
Comment (muglug) [reply src/Composer/Command/ConfigCommand.php]: Couldn't the output of `$input->getArgument()` be something other than a string? It's passed into `preg_match`, which expects one.

File: src/Composer/DependencyResolver/Solver.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (muglug) [reply src/Composer/DependencyResolver/Solver.php]: Oh sorry! Have fixed.

File: src/Composer/Command/BaseCommand.php (hunk 10 above)
Comment (muglug) [reply src/Composer/Command/BaseCommand.php]: This tells Psalm (and PHPStan, and PHPStorm and other type checkers) that the value returned by `getApplication` is a subclass of the return type declared in the parent.

File: src/Composer/Command/ConfigCommand.php (hunk 9 above)
Comment (Seldaek) [reply src/Composer/Command/ConfigCommand.php]: As far I know in this configuration it only returns string or null, which would be falsey anyway. But obviously getArgument can be sometimes an array if the arg is configured as such, so the docs on that have to allow for more return types than are applicable in a given case. You can leave as is I guess it's harmless either way.

File: src/Composer/Command/BaseCommand.php (hunk 10 above)
Comment (Seldaek) [reply src/Composer/Command/BaseCommand.php]: Ahh I see, makes sense thanks.

**Comment by Seldaek** [comment]
Alright, thanks for the updates. I think we might look at Psalm again in the future, maybe with v3 once we can drop old PHP versions and add much more type info inline it'll make it easier to adopt multiple static analyzers.

**Comment by muglug** [comment]
You're welcome! Let me know if you have any questions - I understand the pain of having two competing configs. Some great projects use both, [some great projects use none](https://github.com/sebastianbergmann/phpunit).

YMMV, but there's not much to differentiate Psalm and PHPStan when the type coverage is relatively low (81% for the `src` directory). There only real difference that'll matter to you if/when you attempt to "level up" is configuration - Psalm allows you to pick and choose what issues you care about (and in which files) because we need that level of granularity at Vimeo, whereas PHPStan is a more one-size-fits-all sort of deal.

**Comment by Seldaek** [comment]
Alright thanks for the overview :)
